src/components/templates/Poster/usePosterVideo.ts

```diff
@@ -0,0 +1,132 @@
+  const [room, setRoom] = useState<Room>();
+  const [participants, setParticipants] = useState<
+    Array<LocalParticipant | RemoteParticipant>
+  >([]);
+  const [token, setToken] = useState<string>();
+  const [hasVideo, setHasVideo] = useState(false);
+
+  const { userId } = useUser();
+  const { worldUsersById } = useWorldUsersById();
+
+  const localParticipant = room?.localParticipant;
+
+  useEffect(() => {
+    (async () => {
+      if (!userId) return;
+
+      const getToken = firebase.functions().httpsCallable("video-getToken");
+      const response = await getToken({
+        identity: userId,
+        room: venueId,
+      });
+      setToken(response.data.token);
+    })();
+  }, [venueId, userId]);
+
+  useEffect(() => {
+    if (!localParticipant) return;
+
+    setParticipants((prevParticipants) => [
+      // Hopefully prevents duplicate users in the participant list
+      ...prevParticipants.filter(
+        (p) => p.identity !== localParticipant.identity
+      ),
+      localParticipant,
+    ]);
+  }, [localParticipant]);
+
+  useEffect(() => {
+    if (!token) return;
+
+    const participantConnected = (participant: RemoteParticipant) => {
+      setParticipants((prevParticipants) => [
+        participant,
+        ...prevParticipants.filter((p) => p.identity !== participant.identity),
+      ]);
+    };
+
+    const participantDisconnected = (participant: RemoteParticipant) => {
+      setParticipants((prevParticipants) => {
+        return prevParticipants.filter((p) => p !== participant);
+      });
+    };
+
+    Video.connect(token, {
+      name: venueId,
+      video: hasVideo,
+    }).then((room) => {
+      setRoom(room);
+      room.on("participantConnected", participantConnected);
+      room.on("participantDisconnected", participantDisconnected);
+      room.participants.forEach(participantConnected);
+    });
+  }, [venueId, token, hasVideo]);
+
+  const turnVideoOn = () => {
+    setHasVideo(true);
+  };
+
+  const turnVideoOff = () => {
+    setHasVideo(false);
+  };
+
+  const { passiveListerens, activeParticipants } = useMemo(
+    () =>
+      participants.reduce<{
+        passiveListerens: WithId<User>[];
+        activeParticipants: (RemoteParticipant | LocalParticipant)[];
+      }>(
+        (acc, participant) => {
+          if (participant.videoTracks.size === 0) {
+            return {
+              ...acc,
+              passiveListerens: [
+                ...acc.passiveListerens,
+                withId(
+                  worldUsersById[participant.identity],
+                  participant.identity
+                ),
+              ],
+            };
+          }
+
+          return {
+            ...acc,
+            activeParticipants: [...acc.activeParticipants, participant],
+          };
+        },
+        {
+          passiveListerens: [],
+          activeParticipants: [],
+        }
+      ),
+    [participants, worldUsersById]
+  );
```

Comment:
This.. is confusing and hard to read to say the least. I haven't reviewed the logic/etc of it as I can't even get my brain to comprehend it. I expect it can be rewritten to be far easier to comprehend, perhaps by extracting some of the helper functions, or the reducer function, or similar.

Reply:
I don't know if it's possible to simplify it more than it's done in https://github.com/sparkletown/sparkle/pull/1299/commits/1c71872238b0cf77414c943211ee5416cc65a456 .
If you have any suggestions, please, feel free to unresolve this

Reply:
Looks much better in that diff.

---

What will `participant.videoTracks.size` show when i've 'joined' the video, and then turn my camera off using the controls from my video refactor PR (yet when the user hasn't 'left the video seat')?

Reply:
It will have your videoTrack as well as your camera will be `turned on`

Reply:
> It will have your videoTrack as well as your camera will be turned on

I may just not be understanding what you mean by this, but why will my video be 'turned on' when it is 'hidden' using the controls I mentioned above?